components/brave_wallet_ui/page/screens/composer_ui/select_token_modal/select_token_modal.tsx

```diff
@@ -167,270 +175,362 @@ export const SelectTokenModal = React.forwardRef<HTMLDivElement, Props>(
     )
 
-    const getTokenListWithBalances = React.useCallback(
-      (account: BraveWallet.AccountInfo) => {
-        return getTokenListByAccount(account).filter((token) =>
+    const getAllAccountsWithBalance = React.useCallback(
+      (asset: BraveWallet.BlockchainToken) => {
+        if (!tokenBalancesRegistry) {
+          return []
+        }
+
+        return accounts.filter((account) =>
           new Amount(
-            getBalance(account.accountId, token, tokenBalancesRegistry)
+            getBalance(account.accountId, asset, tokenBalancesRegistry)
           ).gt(0)
         )
       },
-      [getTokenListByAccount, tokenBalancesRegistry]
+      [tokenBalancesRegistry, accounts]
     )
 
-    const getTokensBySelectedSendOption = React.useCallback(
-      (account: BraveWallet.AccountInfo) => {
-        if (selectedSendOption === SendPageTabHashes.nft) {
-          return getTokenListWithBalances(account).filter(
-            (token) => token.isErc721 || token.isNft || token.isErc1155
+    // Memos
+    const tokensBySelectedComposerOption = React.useMemo(() => {
+      if (modalType === 'swap') {
+        return fullVisibleTokenList
+          .filter(
+            (token) => !token.isErc721 && !token.isErc1155 && !token.isNft
           )
+          .filter((token) => swapNetworksChainIds.includes(token.chainId))
+      }
+
+      if (selectedSendOption === SendPageTabHashes.nft) {
+        return userVisibleTokensInfo
+          .filter((token) => token.isErc721 || token.isNft || token.isErc1155)
+          .filter((token) => getAllAccountsWithBalance(token).length > 0)
+      }
+
+      return userVisibleTokensInfo
+        .filter((token) => !token.isErc721 && !token.isErc1155 && !token.isNft)
+        .filter((token) => getAllAccountsWithBalance(token).length > 0)
+    }, [
+      modalType,
+      fullVisibleTokenList,
+      userVisibleTokensInfo,
+      swapNetworksChainIds,
+      selectedSendOption,
+      getAllAccountsWithBalance
+    ])
+
+    const userTokenList: UserAssetInfoType[] = React.useMemo(() => {
+      if (!tokenBalancesRegistry) {
+        // wait for balances before computing this list
+        return []
+      }
+      return tokensBySelectedComposerOption.map((asset) => {
+        return {
+          asset,
+          assetBalance:
+            selectedAccountFilter.address === AllAccountsOption.address
+              ? getFullAssetBalance(asset)
+              : getBalance(
+                  selectedAccountFilter.accountId,
+                  asset,
+                  tokenBalancesRegistry
+                )
         }
-        return getTokenListWithBalances(account).filter(
-          (token) => !token.isErc721 && !token.isErc1155 && !token.isNft
-        )
-      },
-      [getTokenListWithBalances, selectedSendOption]
-    )
+      })
+    }, [
+      tokensBySelectedComposerOption,
+      getFullAssetBalance,
+      tokenBalancesRegistry,
+      selectedAccountFilter
+    ])
```

Comment:
nit: could we create a balance lookup object instead of mapping to `UserAssetInfoType`


```suggestion
    const userTokenBalances: Record<string, string> = React.useMemo(() => {
      if (!tokenBalancesRegistry) {
        // wait for balances before computing this list
        return {}
      }

      const balancesMap: Record<string, string> = {}
      for (const asset of tokensBySelectedComposerOption) {
        balancesMap[getAssetIdKey(asset)] =
            selectedAccountFilter.address === AllAccountsOption.address
              ? getFullAssetBalance(asset)
              : getBalance(
                  selectedAccountFilter.accountId,
                  asset,
                  tokenBalancesRegistry
                )
      }

      return balancesMap
    }, [
      tokensBySelectedComposerOption,
      getFullAssetBalance,
      tokenBalancesRegistry,
      selectedAccountFilter
    ])
```

Reply:
Unless this is `trivial`, I'd prefer not to change this, since we would need to restructure this entire feature.

Reply:
Done!